### PR TITLE
[RISCV][lld] Set the type of TLSDESC relocation's referenced local symbol to STT_NOTYPE

### PR DESCRIPTION
**Pull request by ilovepi**
When adding fixups for RISCV_TLSDESC_ADD_LO and RISCV_TLSDESC_LOAD_LO,
the local label added for RISCV TLSDESC relocations have STT_TLS set,
which is incorrect. Instead, these labels should have `STT_NOTYPE`.

This patch stops adding such fixups and avoid setting the STT_TLS on
these symbols. Failing to do so can cause LLD to emit an error `has an
STT_TLS symbol but doesn't have an SHF_TLS section`. We additionally,
adjust how LLD services these relocations to avoid errors with
incompatible relocation and symbol types.

### CODE REVIEW

File: lld/ELF/Relocations.cpp [hunk 1]
```diff
@@ -1280,7 +1280,6 @@ static unsigned handleTlsRelocation(RelType type, Symbol &sym,
   if (config->emachine == EM_MIPS)
     return handleMipsTlsRelocation(type, sym, c, offset, addend, expr);
   bool isRISCV = config->emachine == EM_RISCV;
-
```
Comment (ilovepi) [inline lld/ELF/Relocations.cpp:1283]: will remove whitespace diff.

File: lld/test/ELF/riscv-tlsdesc.s [hunk 2]
```diff
@@ -30,10 +30,10 @@
 # RUN: llvm-objdump --no-show-raw-insn -M no-aliases -h -d a.32.ie | FileCheck %s --check-prefix=IE32
 
 # RUN: llvm-mc -triple=riscv64 -filetype=obj d.s -o d.64.o
-# RUN: not ld.lld -shared -soname=d.64.so -o d.64.so d.64.o 2>&1 | FileCheck %s --check-prefix=BADTLSLABEL
+# RUN: ld.lld -shared -soname=d.64.so -o d.64.so d.64.o 2>&1 | FileCheck %s --check-prefix=BADTLSLABEL --allow-empty
```
Comment (MaskRay) [inline lld/test/ELF/riscv-tlsdesc.s:33]: --allow-empty is an anti-pattern. Avoid.

Replace `2>&1 | ...` with --fatal-warnings

File: lld/test/ELF/riscv-tlsdesc.s [hunk 3]
```diff
@@ -136,28 +140,30 @@
...
 # IE32-NEXT:         addi    zero, zero, 0x0
 # IE32-NEXT:         addi    zero, zero, 0x0
 # IE32-NEXT:  111cc: auipc   a0, 0x1
 # IE32-NEXT:         lw      a0, 0x80(a0)
 # IE32-NEXT:         add     a0, a0, tp
 
-## FIXME This should not pass, but the code MC layer needs a fix to prevent this.
-# BADTLSLABEL: error: d.{{.*}}.o has an STT_TLS symbol but doesn't have an SHF_TLS section
+# At one point the local TLSDESC labels would be marked STT_TLS, so make sure we don't regress
+# BADTLSLABEL-NOT: error: d.{{.*}}.o has an STT_TLS symbol but doesn't have an SHF_TLS section
```
Comment (MaskRay) [inline lld/test/ELF/riscv-tlsdesc.s:166]: Delete the NOT pattern. If the diagnostic changes, the NOT pattern may likely not get updated.

`# ` => `## Before #85817`

File: lld/ELF/Relocations.cpp [hunk 4]
```diff
@@ -1480,7 +1480,13 @@ template <class ELFT, class RelTy> void RelocationScanner::scanOne(RelTy *&i) {
 
   // Process TLS relocations, including TLS optimizations. Note that
   // R_TPREL and R_TPREL_NEG relocations are resolved in processAux.
-  if (sym.isTls()) {
+  if (sym.isTls() ||
```
Comment (MaskRay) [inline lld/ELF/Relocations.cpp:1483]: Attach this diff
```
-  if (sym.isTls() ||
-      // These RISCV TLSDESC relocations reference a local symbol that won't be
-      // a TLS symbol, but we need to process them in handleTlsRelocation the
-      // same as other TLS relocations.
-      (config->emachine == EM_RISCV &&
-       (type == R_RISCV_TLSDESC_CALL || type == R_RISCV_TLSDESC_LOAD_LO12 ||
-        type == R_RISCV_TLSDESC_ADD_LO12))) {
+  //
+  // Some RISCV TLSDESC relocations reference a local NOTYPE symbol, but we
+  // need to process them in handleTlsRelocation.
+  if (sym.isTls() || oneof<R_TLSDESC_PC, R_TLSDESC_CALL>(expr)) {
```

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
The title is not descriptive. This is primarily a RISC-V side MC issue. Perhaps

`[RISCV,lld] Set the type of TLSDESC relocations referenced local symbol to STT_NOTYPE`

File: lld/test/ELF/riscv-tlsdesc.s (hunk 2 above)
Comment (ilovepi) [reply lld/test/ELF/riscv-tlsdesc.s:33]: Oh, I wasn't aware. Thanks. Do you think we should update the FileCheck.rst to note that other options/patterns are preferred ... or maybe in the TestingGuide.rst?

**Comment by ilovepi** [comment]
Just realized that I don't think we backported this fix.

/cherry-pick #85817

**Comment by ilovepi** [comment]
\cherry-pick #85816, #85817

**Comment by tstellar** [comment]
> \cherry-pick #85816, #85817

You need to pass in the commit hashes. `/commit <sha> `

**Comment by ilovepi** [comment]
/cherry-pick f6f474c4ef9694a4ca8f08d59fd112c250fb9c73, dfe4ca9

**Comment by ilovepi** [comment]
/cherry-pick f6f474c4ef9694a4ca8f08d59fd112c250fb9c73 dfe4ca9b7f4a422500d78280dc5eefd1979939e6

**Pull request merged**